Homework5/SparseMatrix/Entities/SparseMatrix.cs

```diff
@@ -0,0 +1,127 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace SparseMatrixOperations.Entities
+{
+    public class SparseMatrix : IEnumerable<int>
+    {
+        private Dictionary<(int Column, int Row), int> MatrixElements = new Dictionary<(int, int), int>();
+        private (int Columns, int Rows) Size;
+
+
+        public SparseMatrix(int columns, int rows)
+        {
+            if (columns > 0 && rows > 0)
+            {
+                Size = (columns, rows);
+            }
+            else
+            {
+                throw new ArgumentException("Size of matrix isn't appropriate.");
+            }
+        }
+
+        public int this[int col, int row]
+        {
+            get
+            {
+                InputDimensionCheck(col, row);
+
+                if (IsZeroElement(col, row)) // case when there is no element in such indexes, so the value is zero
+                {
+                    return default(int);
+                }
+
+                var elementIndex = (col, row);
+                return MatrixElements.First(e => e.Key.Equals(elementIndex)).Value; // returning the value at corresponding indexes
+            }
+            set
+            {
+                InputDimensionCheck(col, row);
+
+                var elementIndex = (col, row);
+                MatrixElements.Add(elementIndex, value);
```

Comment:
This setter should not store 0 values.
Assume we have a sparse matrix 1 mln x 1mln elements and it has single non-zero element.
Current code will take memory of size 10^12 bytes * N, where N is size of 1 element.

Reply:
Also, consider the case, when element with the same index updates a few times:
```
var sparseMatrix  = new SparseMatrix();
sparseMatrix[0,0] = 0;
sparseMatrix[0,0] = 1;
```